src/utils/graph/routing.js

```diff
@@ -0,0 +1,250 @@
+import {
+  compare,
+  distance1d,
+  angle,
+  nearestOnLine,
+  groupByRow,
+  nodeLeft,
+  nodeRight,
+  nodeTop,
+  nodeBottom
+} from './common';
+
+/**
+ * Finds positions for the given edges relative to their nodes.
+ * Input nodes and edges are updated in-place.
+ * Results are stored in the `points` property on edges.
+ * @param {object} params The layout parameters
+ * @param {array} params.nodes The input nodes
+ * @param {array} params.edges The input edges
+ * @param {number} params.spaceX The minimum gap between a node and passing edges in X
+ * @param {number} params.spaceY The minimum gap between a node and passing edges in Y
+ * @param {number} params.minPassageGap The minimum gap between two nodes in which an edge can pass in X
+ * @param {number} params.stemUnit The unit length for edge stems at anchors
+ * @param {number} params.stemMax The maximum length of edge stems at anchors
+ * @param {number} params.stemMinSource The minimum length for edge stems at source anchors
+ * @param {number} params.stemMinTarget The minimum length for edge stems at target anchors
+ * @param {number} params.stemSpaceSource The ideal spacing between edge stems at source anchors
+ * @param {number} params.stemSpaceTarget The ideal spacing between edge stems at target anchors
+ * @returns {void}
+ */
+export const routing = ({
+  nodes,
+  edges,
+  spaceX,
+  spaceY,
+  minPassageGap,
+  stemUnit,
+  stemMinSource,
+  stemMinTarget,
+  stemMax,
+  stemSpaceSource,
+  stemSpaceTarget
+}) => {
+  // Find the rows formed by nodes
+  const rows = groupByRow(nodes);
+
+  // For each node
+  for (const node of nodes) {
+    // Sort the node's target edges by the angle between source and target nodes
+    // Break ties by alternating increasing X offsets per row
+    node.targets.sort((a, b) =>
+      compare(
+        Math.atan2(
+          b.sourceNode.y - b.targetNode.y,
+          b.sourceNode.x - b.targetNode.x !== 0
+            ? b.sourceNode.x - b.targetNode.x
+            : (b.targetNode.row % 2 === 0 ? -1 : 1) *
+                Math.pow(b.targetNode.row - b.sourceNode.row, 3)
```

Comment:
Is it possible to simplify this? Maybe add some variables? Same for the other one below.

Reply:
I've actually simplified this and removed the extra steps, since it looks like they are not all that effective, so this one should now be fine I think.